### ISSUE TITLE
Getting started - basic network won't classify two distinct patterns

### ISSUE THREAD

**Issue by gschmidt958** [cbovar/ConvNetSharp#154]
Hello,
I'm trying to get to a working starting point and work up from there so I set up a basic net.   I then created a simple problem, two test cases each initialized with a different constant value.  However after running through 1000 iterations of training on both patterns, when I forward the patterns back through the net, I get exactly the same classification probabilities for both patterns (increasing the number of iterations doesn't help).  Initially, I had a single class, but then changed it to two (thinking that might somehow help, but of course it didn't). I'm sure I must be doing something very fundamentally wrong.   If you would be willing to take a look at my code (below), that would be much appreciated.   I'm sure it must be something very basic that's wrong.  Thanks!

Output

![image](https://user-images.githubusercontent.com/5486583/80557400-ccb0c480-89a4-11ea-8d26-c0a7222b4ac7.png)

```csharp
using System;
using ConvNetSharp.Core;
using ConvNetSharp.Core.Layers.Double;
using ConvNetSharp.Core.Training.Double;
using ConvNetSharp.Volume;
using ConvNetSharp.Volume.Double;

namespace ClusterNN
{
internal class Program
{
private static void Main()
{
var net = new Net();

        net.AddLayer(new InputLayer(7, 9, 6));

        // Convolution layer #1 2x2
        net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
        // Result after applying filter is (6 x 8) x 6_filters = 288 elements

        // declare a ReLU (rectified linear unit non-linearity)
        net.AddLayer(new ReluLayer());

        // Convolution layer #2 2x2
        net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
        // Result after applying filter is (5 x 7) x 6_filters = 210 elements

        // declare a ReLU (rectified linear unit non-linearity)
        net.AddLayer(new ReluLayer());

        // Fully connected layer #1.
        net.AddLayer(new FullyConnLayer(100));
        
        // Fully connected layer #2
        net.AddLayer(new FullyConnLayer(2));

        // declare the linear classifier on top of the previous hidden layer
        net.AddLayer(new SoftmaxLayer(2));

        // Construct a trainer for our network
        var trainer = new SgdTrainer(net) { LearningRate = 0.001, Momentum = 0.0, BatchSize = 1 };

        var evenInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));
        var oddInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));

        MakeEvenOddVolumes(evenInput, oddInput);

        var label0_2 = BuilderInstance.Volume.SameAs(new Shape(1, 1, 2));
        label0_2.Set(0, 0, 0, 0.0);
        label0_2.Set(0, 0, 1, 1.0);

        var label1_2 = BuilderInstance.Volume.SameAs(new Shape(1, 1, 2));
        label1_2.Set(0, 0, 0, 1.0);
        label1_2.Set(0, 0, 1, 0.0);

        for (int i = 0; i < 1000; i++)
        {
            Console.WriteLine("iteration: " + i);

            trainer.Train(evenInput, label0_2);
            trainer.Train(oddInput, label1_2);
        }

        // forward a data point through the network
        var vscore = net.Forward(evenInput);
        Console.WriteLine("probability that x is even: " + vscore.Get(0, 0, 0) + " " + vscore.Get(0, 0, 1));

        var vscore2 = net.Forward(oddInput);
        Console.WriteLine("probability that x is odd: " + vscore2.Get(0, 0, 0) + " " +  vscore2.Get(0, 0, 1));

    }

    private static void MakeEvenOddVolumes(Volume<double> evenVolume, Volume<double> oddVolume)
    {
        int counter = 0;
        for (int i = 0; i < 7; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                for (int k = 0; k < 6; k++)
                {
                    evenVolume.Set(i, j, k, 100.0);
                    oddVolume.Set(i, j, k, 11);
                    counter++;
                }
            }
        }
    }
}
```

**Comment by cbovar** [comment]
I noticed the loss is not changing much from one iteration to the other so I played with the learning rate:
Increasing learning rate to `1.0` seems to solve this issue.
Another way would be changing `SgdTrainer `by `AdamTrainer`.

![image](https://user-images.githubusercontent.com/5098385/80557737-81310180-8a0a-11ea-8e00-68cc0dd6a0ac.png)

I'm not sure why such a high learning rate is needed. It means that the computed gradients are very small.

**Comment by gschmidt958** [comment]
Changing the learning rate to 1.0 gives the same result and doesn't appear to solve the problem.  The output for both lines is identical indicating that forwarding two different patterns results in exactly the same classification.  Notice that in the code below, both classifications are output in the same order.  So one would ideally expect as output:
"probability that x is even: 0, 1"
"probability that x is odd: 1, 0"
(correct classification)

But instead we essentially get:
"probability that x is even: 1, 0"
"probability that x is odd: 1, 0"
(no classification at all)

Code:
        // forward a data point through the network
        var vscore = net.Forward(evenInput);
        Console.WriteLine("probability that x is even: " + vscore.Get(0, 0, 0) + " " + vscore.Get(0, 0, 1));

        var vscore2 = net.Forward(oddInput);
        Console.WriteLine("probability that x is odd: " + vscore2.Get(0, 0, 0) + " " +  vscore2.Get(0, 0, 1));

Thanks.

**Comment by cbovar** [comment]
Ah yes, you are right. The classification is indeed incorrect.

I haven't got to the bottom of this yet but by normalizing the input (100.0 -> 1.0, 11.0 -> 0.11) and using `TanhLayer` instead of `ReluLayer` the training converge to a correct classificaiton.

![image](https://user-images.githubusercontent.com/5098385/80562200-96615c80-8a19-11ea-9293-5529348ee5bd.png)

Have you by any chance tried to do the same classification using another framework (e.g. pytorch) ?

**Comment by cbovar** [comment]
Same thing in pytorch:
```python 
import torch
import torch.nn as nn
import numpy as np

class Net(nn.Module):
   def __init__(self):
       super(Net, self).__init__()
       self.conv1 = nn.Conv2d(6, 6, (2,2))
       self.conv2 = nn.Conv2d(6, 6, (2,2))
       self.fc1 = nn.Linear(210, 100)
       self.fc2 = nn.Linear(100, 2)

   def forward(self, x):
       x = nn.functional.relu(self.conv1(x))
       x = nn.functional.relu(self.conv2(x))
       x = torch.flatten(x)
       x = self.fc1(x)
       x = self.fc2(x)
       return x

net = Net()
print(net)

x_even = torch.from_numpy(np.ones(7*9*6).reshape(1,6,9,7) * 100).float()
x_odd = torch.from_numpy(np.ones(7*9*6).reshape(1,6,9,7) * 11).float()

output_even = torch.LongTensor([1])
output_odd = torch.LongTensor([0])

# Define Optimizer and Loss Function
optimizer = torch.optim.SGD(net.parameters(), lr=0.1)
loss_func = torch.nn.CrossEntropyLoss()

for i in range(1000):
    prediction = net(x_even)
    prediction = prediction.unsqueeze(0)
    loss = loss_func(prediction, output_even) 
    optimizer.zero_grad()
    loss.backward()        
    optimizer.step()       

    prediction = net(x_odd)
    prediction = prediction.unsqueeze(0)
    loss = loss_func(prediction, output_odd) 
    optimizer.zero_grad()
    loss.backward()        
    optimizer.step()       

    print(f'iteration #{i} Loss:{loss.data.numpy()}')

vscore = nn.functional.softmax(net(x_even), dim=0)
print(f'probability that x is even: {vscore}')

vscore2 = nn.functional.softmax(net(x_odd), dim=0)
print(f'probability that x is odd: {vscore2}')
```

![image](https://user-images.githubusercontent.com/5098385/80568330-6457f680-8a29-11ea-9fa7-9eb74db995d5.png)


If I normalize the input, we get:
![image](https://user-images.githubusercontent.com/5098385/80568434-91a4a480-8a29-11ea-8b80-da5216ccea6e.png)

Si I guess what we learn is that we should normalize inputs (in ConvNetSharp and Pytorch and probably any other deep learning library)

**Comment by gschmidt958** [comment]
Hi,
Thank you for looking into this.  I'm coming at this from an existing C# program which I'm trying to enhance via use of a ConvNet, so I have not been using pytorch (I'm more of a C# guy which is why I'm here in the first place :) ).

My initial toy problem (even/odd) was normalized in the sense that it was given 0/1 (binary) input.  Since that didn't initially work, I erroneously though that trying numbers with greater magnitude as input might help (hence the arbitrary 100, 11), but as you pointed out, normalized input is best.  So given what I know now, there are two remaining questions:

1) Your intuition told you that Tanh should be used instead of RELU and so I wonder why RELU doesn't work here.  Do you think the problem might be the "dying RELU" problem?  (just guessing here).  Or maybe RELU either works great when it works, or else doesn't work at all for some particular reason.

2) Since my problem is to produce an evaluation function in the range {0,1}, I really only require a single probability for output.  So when I change the output layer (fully connected + softmax) back to a single output representing a probability (single class), and train even/odd with 0/1 training set, it no longer works.  When forwarding a pattern, I would expect a probability indicating the degree to which the input corresponds to and even or odd pattern, but it doesn't.  Instead, I'm back to getting exactly 1 for both patterns. 
 Shouldn't that work, or am I missing something fundamental here? (code below)

Observation: It seems as though constructing even a working network may be a bit of a "black art".

Thanks!

```csharp
namespace ClusterNN
{
    internal class Program
    {
        private static void Main()
        {
             net.AddLayer(new InputLayer(7, 9, 6));

            // Convolution layer #1 2x2
            net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
            // Result after applying filter is (6 x 8) x 6_filters = 288 elements

            // declare a ReLU (rectified linear unit non-linearity)
            net.AddLayer(new TanhLayer());

            // Convolution layer #2 2x2
            net.AddLayer(new ConvLayer(2, 2, 6) { Stride = 1 });
            // Result after applying filter is (5 x 7) x 6_filters = 210 elements

            // declare a ReLU (rectified linear unit non-linearity)
            net.AddLayer(new TanhLayer());

            // Fully connected layer #1.
            net.AddLayer(new FullyConnLayer(100));
            
            // Fully connected layer #2
            net.AddLayer(new FullyConnLayer(1));

            // declare the linear classifier on top of the previous hidden layer
            net.AddLayer(new SoftmaxLayer(1));

            // Construct a trainer for our network
            var trainer = new SgdTrainer(net) { LearningRate = .001, Momentum = 0.0, BatchSize = 1 };

            var evenInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));
            var oddInput = BuilderInstance.Volume.SameAs(new Shape(7, 9, 6));

            MakeEvenOddVolumes(evenInput, oddInput);

            var label0 = BuilderInstance.Volume.SameAs(new Shape(1, 1, 1));
            label0.Set(0, 0, 0.0);

            var label1 = BuilderInstance.Volume.SameAs(new Shape(1, 1, 1));
            label1.Set(0, 0, 1.0);

            for (int i = 0; i < 1000; i++)
            {
                Console.WriteLine("iteration: " + i);

                trainer.Train(evenInput, label0);
                trainer.Train(oddInput, label1);
            }

            // forward a data point through the network
            var vscore = net.Forward(evenInput);
            Console.WriteLine("probability that x is even: " + vscore.Get(0, 0, 0));

            var vscore2 = net.Forward(oddInput);
            Console.WriteLine("probability that x is odd: " + vscore2.Get(0, 0, 0));

        }

        private static void MakeEvenOddVolumes(Volume<double> evenVolume, Volume<double> oddVolume)
        {
            int counter = 0;
            for (int i = 0; i < 7; i++)
            {
                for (int j = 0; j < 9; j++)
                {
                    for (int k = 0; k < 6; k++)
                    {
                        evenVolume.Set(i, j, k, 1.0);
                        oddVolume.Set(i, j, k, .11);
                        counter++;
                    }
                }
            }
        }
    }
}
```

**Comment by cbovar** [comment]
1. It should work with Relu / LeakyRelu as well. Initially I noticed that the gradients of even sample where much bigger than odd sample so I thought of using a tanh or sigmoid layer to prevent big values flowing into the network. 

2. If you want to produce a single output value, then you should use a `RegressionLayer` as last layer rather than `SoftmaxLayer`:

![image](https://user-images.githubusercontent.com/5098385/80663549-e3e7d300-8ac6-11ea-88f5-20cf0e3edcca.png)

**Comment by gschmidt958** [comment]
Thank you so much for your awesome support, I think I've reached a point where I can make real progress now and now I'm really looking forward to that!
